src/crimson/os/seastore/transaction_manager.h

```diff
@@ -549,7 +549,8 @@ class TransactionManager : public SegmentCleaner::ExtentCallbackInterface {
       device_type_t::SEGMENTED,
       std::make_unique<SegmentedAllocator>(
 	*segment_cleaner,
-	*sm));
+	*sm,
+	journal->get_segment_seq_allocator()));
```

Comment:
I think this PR is to revert https://github.com/ceph/ceph/pull/44809/commits/1446c58acd85c4c8264f4a38c23be36dc2761691 that has decoupled `EPM::Writer` from `Journal` by assigning a static `OOL_SEG_SEQ` to all the out-of-line segments, instead of the segment sequence from `Journal`. Because I thought the segment sequence from `Journal` may not be consistent to the journal segment sequence that will be rolled and committed later. And IIUC setting to `OOL_SEG_SEQ` was simpler and didn't break the relay.

> because we rely on this to judge whether a delta needs to be replayed

Would you be able to point out where is the code to do the judgement?

And I don't quite follow this problem, as the existing unit tests have already verified the replaying process and the replayed results, right?

Reply:
When I run the unittest for omap manager, an error of replaying an outdated delta came up. The judgement is [here](https://github.com/ceph/ceph/blob/a43b68294cfef9ca0ffe72b8401935c4f3ba567b/src/crimson/os/seastore/journal/segmented_journal.cc#L212), the comments just above that line of code explained why the judge matters.

The scenario we are trying to deal with in this pr is like this:

1. an extent `E` is written to an ool segment `S`;
2. an modification to `E` happens, which creats a delta `D`
3. segment `S` gets reclaimed, `E` is rewritten somewhere else
4. segment `S` gets opened, and the address where `E` stays originally gets written with other data
5. crimson-osd restart

In this scenario, delta `D` shouldn't be replayed, because its effect is already merged to `E` when `E` gets rewritten; but the current journal still replays `D`, which should have been prevented by comparing `D`'s segment seq and segment `S`'s segment seq

Reply:
Thank you, I think this scenario makes sense to me, this is really an issue to apply a delta to an extent that has been overwritten.

But I'm still wondering whether this PR really addresses the issue, consider the following case:
1. an extent `E` is written to an ool segment `S`, with **journal segment sequence** `A`
2. an modification to `E` happens, which creates a delta `D`, submitted to the journal segment at sequence `A`
3. segment `S` gets reclaimed, `E` is rewritten somewhere else, while the current journal segment is still at sequence `A`
4. segment `S` gets re-opened, and the physical location of `E` gets overritten with new data. However, the current journal segment is still at sequence `A`, meaning the re-opened segment `S` still sets **journal segment sequence** `A`

The above might look rare right now, but I believe the possibility still exists as we could split more and more data to out-of-line segments.

I think the real problem is that the out-of-line segments are opened and reclaimed independently regarding to journal segments, ool segments might need to maintain an independent sequence in order to (somehow) detect if it is outdated during replay in a reliable way.

Reply:
you're right. I wonder if we can solve this by assigning the next **journal segment sequence** to ool segments when opening them, just like what we do when opening journal segments?

Reply:
you're right, I think we can give ool segments an independent sequence, and record the ool segments' segment_seq in delta_info_t, so we can see whether the extents corresponding to a delta_info_t has rewritten.

What do you think, thanks:-)

Reply:
>  for journal segments, we only keep to guarantee the segment sequence strictly monotonously increase, and no need to keep journal segments' segment sequences continuous.

Hmm, I don't see any issues to relax the constraint by allowing the journal segment sequence to be non-consecutive. @athanatos do you have any concerns?

Reply:
> Hmm, I don't see any issues to relax the constraint by allowing the journal segment sequence to be non-consecutive. @athanatos do you have any concerns?

@cyx1231st  Oh, sorry, that comment is not correct  and I've deleted it. The main problem is that it would prevent deltas that need to be replayed from being replayed, for example:

1. journal segment `A` is opened with segment seq `N`
2. ool segment `B` is opened with segment seq `N+1`
3. delta `D` corresponds to extents in `B` are recorded in journal segment `A` 

In this scenario, delta `D` should be replayed, but it won't, because `B`'s segment seq is larger than that of delta `D`'s.

> you're right, I think we can give ool segments an independent sequence, and record the ool segments' segment_seq in delta_info_t, so we can see whether the extents corresponding to a delta_info_t has rewritten.

@cyx1231st @athanatos What do you think about this method? Thanks:-)

Reply:
I think that writing the corresponding paddr's segment seq into `delta_info_t`, and during replay, check whether the sequence matches seems working. Let's see whether it really works.